french3.csl

```diff
@@ -208,9 +208,11 @@
       <text macro="title" suffix=". "/>
       <choose>
         <if type="thesis">
-          <text variable="genre" suffix=". "/>
-          <text macro="publisher" prefix=" "/>
-          <text variable="number-of-pages" suffix=" p."/>
+          <group delimiter=" ">
+            <text variable="genre" suffix="."/>
+            <text macro="publisher"/>
```

Comment:
technically you're adding a space here--but if you look up at the publisher macro, it's a huge mess anyway--so unless you're super motivated, I'd just leave this as you have it here. If you want to go all out, the publisher macro would have to all go into a group with delimiters and have trailing spaces removed from affixes.